source/common/init/target_impl.cc

```diff
@@ -45,8 +45,14 @@ bool TargetImpl::ready() {
   if (watcher_handle_) {
     // If we have a handle for the ManagerImpl's watcher, signal it and then reset so it can't be
     // accidentally signaled again.
+    // NOTE: The still_alive_ guard is used here to avoid the scenario in which as a result of
+    // calling ready() this target is destroyed. This is possible in practice, for example when
+    // a listener is deleted as a result of a failure in the context of the ready() call.
+    std::weak_ptr<bool> weak_still_alive = still_alive_;
     const bool result = watcher_handle_->ready();
-    watcher_handle_.reset();
+    if (!weak_still_alive.expired()) {
```

Comment:
should this still_alive be added and checked by shared_target? Or we can wait until the next crash :)
```SharedTargetImpl::ready()```

Reply:
Yeah, if we agree this is a legitimate fix, I will fix this also and add specific tests for this. Do we believe this is the right fix though? IMO it is, as this is the crashing flow I as hitting: listener starts warming -> listener warm completes and calls ready on listener manager -> listener manager does final init of listener including calling listen() -> listen() fails -> listener manager removes listener -> crash. This is all in the same call stack.

Reply:
I did a slightly different fix after talking with @htuch. I also did it in both places. PTAL.

Reply:
New solution looks good modulo nits.